fix(1467, 1623): improve number inputs

shubha-rajan opened the pull request:
### Description
A continuation of work in #1647 to improve user experience for number inputs by
- Adding `input=decimal` to use a more accessible keyboard on iOS (#1467 )
- Using built-in html validation of number type to enforce that numbers are integers, and convert any float inputs (if they somehow get past client-side validation) to `Long` on the back end (#1623)
- Validate min and max client-side using built-in html validation before submitting

### Screenshots:
On iOS 12 before adding decimal input:
<img src="https://user-images.githubusercontent.com/19631367/142085969-c75b9c31-d737-4e5d-984d-c216cb7a7651.png" alt="screenshot-without-decimal-input" width="400"/>
On iOS 12 after adding decimal input:
<img src="https://user-images.githubusercontent.com/19631367/142086275-2cfd0d9e-2845-4b08-a692-216cc74f4a33.png" alt="screenshot-with-decimal-input" width="400"/>
Validation for integer input: 

<img width="1040" alt="Screen Shot 2021-11-12 at 3 59 24 PM" src="https://user-images.githubusercontent.com/19631367/142086396-57816ac6-4353-4fcc-9728-048151461a74.png">

Validation for min/max:
![image](https://user-images.githubusercontent.com/19631367/142092779-0b0da8a5-4cd6-47ee-9376-b354d57e255e.png)

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #1467; Fixes #1623;

shanemc-goog commented on universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:
```diff
@@ -376,7 +376,13 @@ private void putParentArray(Path path) {
     } catch (JsonPathTypeMismatchException e) {
-      return Optional.empty();
+      try {
+        // check if value is a double/float and cast to long if needed
```
The Question type definition is to only allow integers, and the FE tries to enforce that, so I think we should have consistency with the backend checking too.

shanemc-goog replied:
This still needs reverting correct?

shanemc-goog replied:
?

shanemc-goog commented on universal-application-tool-0.0.1/app/assets/javascripts/number.ts:
```diff
@@ -8,8 +8,9 @@ class NumberController {
-        const n = parseInt(numberQuestion.value.replace(/[^.\d]/g, ''));
+        // Block non-numeric characters like like 'e', 'E', '+', '-'
+        // which will cause problems when parsing to a number
+        const n = parseFloat(numberQuestion.value.replace(/[^.\d]/g, ''));
```
I think we probably want a validation.ts method to show an error message rather than manipulating user key strokes. This is similar to currency and Address Zip code validation where we validate upon Submit.

shanemc-goog replied:
app/assets/javascripts/validation.ts

shanemc-goog replied:
Do we still need this?  All the validation should be a submit not typing time.

shanemc-goog replied:
I think we still don't need this BTW.

shanemc-goog commented:
I left some high level comments, but in general the Number type is defined as an Integer as it, so we should keep it that way without a broader discussion to redefine it, or differentiate Int vs float.

go/uat-question-requirements -> 
https://docs.google.com/document/d/1HYVHBno6dk7fmS5w6V6Y3jxERHGfDoQKTtzBZr27Y6o/edit#

shubha-rajan commented:
Updatied this to use the same onSubmit validation method as the other question types in `validation.ts`. Invalid input now looks like this (feel free to suggest a better error message): 
![image](https://user-images.githubusercontent.com/19631367/142335091-514d9123-672f-4640-8b7c-ff69ca80c552.png)

This PR still relies on the built-in HTML validation for the number input type to enforce min and max instead of rendering those errors in the error div. Would it be better to write our own client-side validation?

shanemc-goog commented on browser-test/src/number_input.test.ts:
```diff
@@ -1,9 +1,27 @@
 } from './support'
 
+describe('input type', () => {
+  it('sets the inputmode attribute to decimal', async () => {
```
I don't think we need this test as it is more of a "change detector" test (Did the attributes change?) where we should be testing functionality as you have below.

shubha-rajan replied:
removed this and added a test to `FieldWithLabelTest.java` to test for this

shanemc-goog commented on browser-test/src/number_input.test.ts:
```diff
@@ -27,16 +45,14 @@ describe('input validation for number questions', () => {
-      '123', 'abc123', '123abc', '12!@#$%^&*()3', '12[]3', '12d3', '12e3', '12E3', '12+3', '12-3'
-    ]
-    const expectedValue = '123';
+    const testValues = ['12e3', '12E3', '-123']
```
include a decimal value

shanemc-goog commented on universal-application-tool-0.0.1/app/views/components/FieldWithLabel.java:
```diff
@@ -153,6 +156,25 @@ public FieldWithLabel setAttribute(String attribute) {
+  public FieldWithLabel setMin(OptionalLong value) {
+    if (!this.fieldType.equals("number")) {
+      throw new RuntimeException(
+          "setting an OptionalLong min value is only available on fields of type `number`");
```
Change the tics around number to single quotes, here and below.

shubha-rajan replied:
done

shanemc-goog commented on universal-application-tool-0.0.1/app/views/components/FieldWithLabel.java:
```diff
@@ -241,6 +263,17 @@ public ContainerTag getContainer() {
+
+      // Set min and max values for client-side validation
+      if (this.minValue.isPresent()) {
+        fieldTag.attr("min", minValue.getAsLong());
```
May want to add a simple test that these are set in the output in test/views/components/FieldWithLabelTest.java

shubha-rajan replied:
done

shanemc-goog commented:
> Updated this to use the same onSubmit validation method as the other question types in `validation.ts`. Invalid input now looks like this (feel free to suggest a better error message): ![image](https://user-images.githubusercontent.com/19631367/142335091-514d9123-672f-4640-8b7c-ff69ca80c552.png)
> 
> This PR still relies on the built-in HTML validation for the number input type to enforce min and max instead of rendering those errors in the error div. Would it be better to write our own client-side validation?

I don't really have an opinion, so whatever looks better to you.

shanemc-goog commented:
Looking good!

shanemc-goog commented on universal-application-tool-0.0.1/conf/messages.en-US:
```diff
@@ -71,6 +71,7 @@ validation.textTooShort=Must contain at least {0} characters.
 validation.textTooLong=Must contain at most {0} characters.
 validation.numberTooSmall=Must be at least {0}.
 validation.numberTooBig=Must be at most {0}.
+validation.numberNonInteger=Number must be a positive whole number and can only contain numeric characters 0-9.
```
This needs to be duplicated in conf/messages

shubha-rajan replied:
done

shubha-rajan commented on browser-test/src/number_input.test.ts:
```diff
@@ -27,16 +45,14 @@ describe('input validation for number questions', () => {
-      expect([testValue, await page.inputValue(numberInput)])
-        .toStrictEqual([testValue, expectedValue]);
+      await applicant.clickNext();
+      expect([testValue, await page.isHidden(numberInputError)]).toStrictEqual([testValue, false]);
```
exactly!

shanemc-goog commented:
Just wanted to make sure you saw the comments in number.ts and ApplicantData.java

shubha-rajan commented:
> Just wanted to make sure you saw the comments in number.ts and ApplicantData.java

I reverted the changes in ApplicantData.java and removed number.ts (moved the validation logic to validation.ts instead) so I can't see the comments anymore.
